### ISSUE TITLE
`upload.saveFilePart` RangeError: Source is too large         at Uint8Array.set

### ISSUE THREAD

**Issue by phucpnt** [alik0211/mtproto-core#105]
Hi,

Could you please let me know reason why we have 4 extra bytes for sending the file parts as the following code:

https://github.com/alik0211/mtproto-core/blob/f19e0c60d7aefab52b9f3977dd67fd7ba60e9ddc/src/tl/serializer/index.js#L58

It cause the issue not able to sending file to telegram. Could you please suggest any way to work around this issue?

**Comment by alik0211** [comment]
Hi. Please show us the code that causes the error.

**Comment by a1im** [comment]
Hi,

Same problem, here's the code

```
const uploadFile = async (bytes, name = '') => {
    const size = bytes.byteLength;
    const partSize = 1024 * 512; // 512kb
    const bigFileLimit = 1024 * 1024 * 10; // 10mb
    const parts = Math.ceil(size / partSize);
    const isBig = size >= bigFileLimit;
    const id = await genMsgId();
    const inputFile = {
        _: isBig ? 'inputFileBig' : 'inputFile',
        parts,
        name,
        id,
    };
    const b = new Uint8Array(bytes);
    let offset = 0;
    let i = 0;

    while (offset !== size) {
        const end = Math.min(size, offset + partSize);
        const splitted = b.slice(offset, end);
        const response = await TgUserApi.call(isBig ? 'upload.saveBigFilePart' : 'upload.saveFilePart', {
            file_id: id,
            file_part: i,
            file_total_parts: parts,
            bytes: splitted,
        });

        offset = end;
        i++;
    }

    return inputFile;
};
```

**Comment by alik0211** [comment]
I'll fix it this weekend. To find out about the new version, follow the official Telegram channel — https://t.me/mtproto_core.

**Comment by azhao12345** [comment]
I tried this, and it looks like you can't send pieces to telegram that are larger than 500k, so you have to break them up and upload the small pieces
see here what pyrogram does: https://github.com/pyrogram/pyrogram/blob/dadb4b4eb68a1af967611adbd08fdc2db599e06e/pyrogram/methods/advanced/save_file.py#L109

here is a sample I made which works
```
async function filize() {
    // slurp the file
    fs.readFile('myfile', async (err, fileData) => {
    try {
        console.log('got the new file');
        console.log(fileData.length);
        console.log(fileData);
        const partId = Math.floor(Math.random() * 999999);
        const partSize = 400 * 1024;
        const totalParts = Math.ceil(fileData.length / partSize);
        console.log(totalParts);
        for (var i = 0; i < totalParts; i++) {
            console.log('starting part');
            const saveFileResult = await mtproto.call('upload.saveFilePart', {
                file_id: partId,
                file_part: i,
                bytes: fileData.slice(i * partSize, Math.min(fileData.length, (i + 1) * partSize))
            });
            console.log('save file result');
            console.log(saveFileResult);
        }
        const sendMediaResult = await mtproto.call('messages.sendMedia', {
            peer: {
                _: 'inputPeerUser',
                user_id: 123718927213
            },
            media: {
                _: 'inputMediaUploadedPhoto',
                file: {
                    _: 'inputFile',
                    id: partId,
                    parts: totalParts,
                    name: 'testfilename.png'
                }
            },
            message: 'test caption',
            random_id: Math.floor(Math.random() * 999999)
        });
        console.log('send media result');
        console.log(sendMediaResult);
    } catch(e) { console.log(e); }
    });
}

```

**Comment by n2j7** [comment]
@azhao12345 there are no chance to work due to file chunk size have to be `524288 % part_size = 0` as described [here](https://core.telegram.org/api/files)

it looks like Serializer buffer can't include 512kb of data of your file chunk
https://github.com/alik0211/mtproto-core/blob/f19e0c60d7aefab52b9f3977dd67fd7ba60e9ddc/src/tl/serializer/index.js#L6

@alik0211 can you add some extra space in buffer for `upload.saveBigFilePart` it looks like 256 bytes is enough
```this.buffer = new ArrayBuffer(524288+256);```

**Comment by alik0211** [comment]
This issue is fixed in version 5.1.0 - https://t.me/mtproto_core/24
